blog/_posts/2023-05-26-scala-3.3.0-released.md

```diff
@@ -0,0 +1,162 @@
+    if n >= 0 then Math.sqrt(n) else break(None)
+  }
+  Some(roots.sum)
+}
+```
+
+When you run the above method, you will notice that it returns None when any of the input elements is negative. Moreover, thanks to `boundary`/`break`, you can see from the console output that it stops the entire process after encountering the first negative element.
+
+`break` can jump out of the`boundary` in the function deeper on the stack. To make it safe, only functions with a matching `Label` in their using clauses can break. This can be used to create isolated parts of an application with streamlined error handling.
+
+Library authors can abstract over `break` and `Label`, creating a nice API for error handling or dealing with uncertain data. We can define the following micro-library as a simplified example:
+
+```Scala
+import util.boundary, boundary.{ Label, break }
+
+type Uncertain[T] = Label[Left[String, Nothing]] ?=> T
+
+def fail(error: String): Uncertain[Nothing] = break(Left(error))
+```
+
+Then it can be used in the following way to create a fragment of a larger application with isolated error handling:
+
+```Scala
+def getUser(id: String): Uncertain[User] =
+  val user = findUser(id)
+  if verifyUser(user) then user else fail("Incorrect user")
+
+def getAddress(user: User): Uncertain[Address] =
+  findAddress(user) match
+    case Some(address) if verifyAddress(address) => address
+    case Some(_) => fail("Incorrect address")
+    case None => fail("Missing address")
+
+def getShippingData(id: String): Either[String, (User, Address)] =
+  boundary {
+    val user = getUser(id)
+    val address = getAddress(user)
+    Right((user, address))
+  }
+```
+
+### The new default implementation of lazy vals
+
+Last but not least, we have changed the default implementation of lazy vals. The new implementation has better performance and is safer under parallel access. This can lead to improvements in the effect systems performance.
+
+## Should I update to Scala 3.3.0?
+
+If you are maintaining a standalone Scala 3 project without external projects depending on it, feel free to switch to Scala 3.3.0 at any moment. You will enjoy all the improvements in the newest version of the compiler.
+
+If you are a library author following semantic versioning, we advise you to update to Scala 3.3.0 in the next minor release of your project. Users of your library would also need to bump their compiler version to use the newest version of your library. If you have already published a minor version on 3.3.0 but learned about a critical vulnerability in your library, you can release a new patch release for the previous minor version of your library using Scala 3.2.2 (or any other version of the compiler you were using previously). This patch can be consumed by all the users of previous versions of the library, no matter if they have already switched to 3.3.0 or not.
+
+### Known incompatibility: Stability of inline parameters
```

Comment:
This is also for the "Changes with compatibility concerns" section.

You probably want to make it more explicit that this a bug fix (so OK to do in a minor version), but for which we have already noticed existing problems in the wild, and so we're helping users by proactively giving them more information on how to deal with it.